src/gui/inputdevices/TouchInputHandler.cpp

```diff
@@ -36,22 +38,27 @@ auto TouchInputHandler::handleImpl(InputEvent const& event) -> bool {
     }
 
     if (event.type == MOTION_EVENT && this->primarySequence) {
-        // Only zoom if there are two fingers involved
-        if (this->primarySequence && this->secondarySequence) {
+        if (this->primarySequence && this->secondarySequence && zoomGesturesEnabled) {
             zoomMotion(event);
-        } else {
+        } else if (event.sequence == this->primarySequence) {
             scrollMotion(event);
         }
     }
 
     if (event.type == BUTTON_RELEASE_EVENT) {
         // Only stop zooing if both sequences were active (we were scrolling)
-        if (this->primarySequence != nullptr && this->secondarySequence != nullptr) {
+        if (this->primarySequence != nullptr && this->secondarySequence != nullptr && zoomGesturesEnabled) {
             zoomEnd();
         }
 
         if (event.sequence == this->primarySequence) {
-            this->primarySequence = nullptr;
+            // If secondarySequence is nullptr, this sets primarySequence
+            // to nullptr. If it isn't, then it is now the primary sequence!
+            this->primarySequence = this->secondarySequence;
+            this->secondarySequence = nullptr;
+
+            this->priLastAbs = this->secLastAbs;
+            this->priLastRel = this->secLastRel;
```

Comment:
If I understand this part correctly, I should, if zoom gestures and touch drawing are disabled, be able to scroll with one finger, put a second finger on the screen, release the first finger and continue scrolling with the second finger. But this code doesn't accomplish that, because `secLastAbs` and `secLastRel` are only kept up to date if `zoomGesturesEnabled` is true. I wrote a quick fix for this: https://github.com/eliasrg/xournalpp/commit/b9c50aab62ce58d157b5960f8d9c23dd1f98e7c4 and it does work when I test it. However, this logic is getting a bit confusing/unnatural, so it might be better to restructure the file more drastically...

Reply:
I think `TouchInputHandler::scrollMotion` should already be keeping these up to date:
```cpp
void TouchInputHandler::scrollMotion(InputEvent const& event) {
    // Will only be called if there is a single sequence (zooming handles two sequences)
    auto offset = [&]() {
        auto absolutePoint = utl::Point{event.absoluteX, event.absoluteY};
        if (event.sequence == this->primarySequence) {
            auto offset = absolutePoint - this->priLastAbs;
            this->priLastAbs = absolutePoint;
            return offset;
        } else {
            auto offset = absolutePoint - this->secLastAbs;
            this->secLastAbs = absolutePoint;
            return offset;
        }
    }();
```

I might be misunderstanding this...

Reply:
But `scrollMotion` is not called for the second finger if there are two fingers on the screen, only for the one actually scrolling.

I think you can see the difference if you compile both versions and try switching fingers while scrolling (zoom gestures and touch drawing turned off)?

Reply:
Yes! With touch-zooming off, I am hitting the issue! Thank you for catching this!

I agree with you about re-structuring the file -- it is getting rather difficult to understand what is going on.

Reply:
Okay, I tried to restructure this in a nice way, but I encountered strange bugs, and eventually I realised that I had misunderstood how the touch input works. Because it doesn't happen only in this file. If you comment out the whole of `handleImpl` except `return false`, you can still scroll with one finger without problem. But as soon as you put a second finger on the screen, it stops working. This is because there is a [`GtkScrolledWindow`](https://developer.gnome.org/gtk3/stable/GtkScrolledWindow.html) which has built-in support for one-finger touches, and even provides kinetic scrolling (which the code in `scrollMotion` doesn't, as you can check by scrolling after switching fingers in my "quick fix" https://github.com/eliasrg/xournalpp/commit/b9c50aab62ce58d157b5960f8d9c23dd1f98e7c4).

So I don't know why there are two competing systems for scrolling... it seems like a bad idea unless I am missing something. Can anybody explain the reason for this? My feeling is that we should either only handle zooming here and leave scrolling for the `GtkScrolledWindow` (probably the best), or get rid of the scrolled window and handle _all_ scrolling ourselves (more work and we wouldn't have kinetic scrolling, but it'd play nicer with zooming and multitouch).

Does that make sense?

Reply:
Thank you for looking into this!
The "enable touch workaround" setting switches between the two scrolling methods.

Reply:
It seems like a proper fix for the transition from two-finger panning to one-finger scrolling will require some refactoring... 
Do you think it could be appropriate to make this fix the subject of a follow-up PR?

Reply:
> So I don't know why there are two competing systems for scrolling... it seems like a bad idea unless I am missing something. Can anybody explain the reason for this?

@eliasrg This is actually a fix for a very old problem that was ported to the new input system. When you enable touch drawing (the touch workaround) you were essentially unable to scroll at all. Back in the day this was fixed with this convoluted system of handling scrolling manually in this case. As you mentioned GTK actually would handle this for us but only for "single finger touches" which are used for drawing in the workaround.

I see two solutions here:
1. We stick with this convoluted system and fix it as well as possible. This would keep our code somewhat smaller and leave stuff like kinetic scrolling with GTK. Arguably it would make the input system harder to understand but also minimize potential areas where we need to fix stuff.
2. We throw out GTK scroll handling completely and implement stuff like kinetic scrolling in our own handlers. I would then make it optional if scrolling should be done with two fingers or one (one being default as it is consistent with GTK). This would increase our code size but make the system straight forward.

@Febbe, @idotobi, @rolandlo, @Technius, @personalizedrefrigerator and @eliasrg what is your opinion on this? Lets make a poll with reaction smileys (🎉 for 1 and 🚀 for 2)

Reply:
@LittleHuba If we are okay with not having two-finger scrolling (personally, I don't need it, but if you are writing with your finger you probably do), then I'd say let GTK do all scroll handling and only handle zoom ourselves. If we do need two-finger scrolling, it seems we've outgrown GTK's scrolling and should throw it out! Since we already have a working scroll handler (which I am working on simplifying), the most difficult part to implement would probably be kinetic scrolling, if indeed we need that. We'd also get a slightly nicer feel overall, with smooth transitions between scrolling and zooming.

Reply:
Okay the verdict is quite clear. Well go forward with option two. As this PR is already big enough I'd recommend a follow-up PR for this. @personalizedrefrigerator and @eliasrg would you like to join up and work on this?